src/main/kotlin/com/exactpro/th2/rptdataprovider/grpc/RptDataProviderGrpcHandler.kt

```diff
@@ -0,0 +1,387 @@
+            Counter.build("th2_grpc_single_requests_get", "GRPC single requests get")
+                .register()
+
+        private val singleRequestProcessed: Counter =
+            Counter.build("th2_grpc_single_requests_processed", "GRPC single requests processed")
+                .register()
+
+        private val streamRequestGet: Counter =
+            Counter.build("th2_grpc_stream_requests_get", "GRPC stream requests get")
+                .register()
+
+        private val streamRequestProcessed: Counter =
+            Counter.build("th2_grpc_stream_requests_processed", "GRPC stream requests processed")
+                .register()
+
+        private val logger = KotlinLogging.logger {}
+
+    }
+
+    val cradleService = this.context.cradleService
+    private val eventCache = this.context.eventCache
+    private val messageCache = this.context.messageCache
+    private val jacksonMapper = context.jacksonMapper
+    private val checkRequestAliveDelay = context.configuration.checkRequestsAliveDelay.value.toLong()
+    private val keepAliveTimeout = context.configuration.keepAliveTimeout.value.toLong()
+
+    private val searchEventsHandler = this.context.searchEventsHandler
+    private val searchMessagesHandler = this.context.searchMessagesHandler
+
+    private val eventFiltersPredicateFactory = this.context.eventFiltersPredicateFactory
+    private val messageFiltersPredicateFactory = this.context.messageFiltersPredicateFactory
+
+    private val sseEventSearchStep = this.context.sseEventSearchStep
+
+
+    private suspend fun checkContext(context: io.grpc.Context) {
+        while (coroutineContext.isActive) {
+            if (context.isCancelled)
+                throw ChannelClosedException("Channel is closed")
+
+            delay(checkRequestAliveDelay)
+        }
+    }
+
+
+    private suspend fun keepAlive(
+        writer: StreamWriter,
+        lastScannedObjectInfo: LastScannedObjectInfo,
+        counter: AtomicLong
+    ) {
+        while (coroutineContext.isActive) {
+            writer.write(lastScannedObjectInfo, counter)
+            delay(keepAliveTimeout)
+        }
+    }
+
+    private fun <T> sendErrorCode(responseObserver: StreamObserver<T>, e: Exception, status: Status) {
+        responseObserver.onError(
+            status.withDescription(ExceptionUtils.getRootCauseMessage(e) ?: e.toString()).asRuntimeException()
+        )
+    }
+
+    private fun errorLogging(e: Exception, requestName: String, stringParameters: String, type: String) {
+        logger.error(e) { "unable to handle request '$requestName' with parameters '$stringParameters' - $type" }
+    }
+
+    private fun <T> handleRequest(
+        responseObserver: StreamObserver<T>,
+        requestName: String,
+        useStream: Boolean,
+        request: MessageOrBuilder?,
+        calledFun: suspend () -> Any
+    ) {
+        val stringParameters = request?.let { TextFormat.shortDebugString(request) } ?: ""

```

Comment:
`TextFormat.shortDebugString(request)` is a heavy operation please avoid execute it without necessary

Reply:
You can use `by laze {}` if you need to compute that once